PXC-3722: Merge codership/mysql-wsrep 5.7.34

venkatesh-prasad-v opened the pull request:
https://jira.percona.com/browse/PXC-3722
    
    Merge tag 'wsrep_5.7.34-25.26' into codership_5.7.34_merge
    
    Merge with conflicts.
    
    Conflicts:
            cmake/wsrep.cmake
            man/mysqldumpslow.1
            mysql-test/suite/galera/r/galera_defaults.result
            scripts/wsrep_sst_common.sh
            scripts/wsrep_sst_rsync.sh
            scripts/wsrep_sst_xtrabackup-v2.sh
            sql/binlog.cc
            sql/handler.cc
            sql/log_event.cc
            sql/mdl.cc
            sql/mysqld.cc
            sql/rpl_slave.cc
            sql/sql_class.cc
            sql/sql_class.h
            sql/sql_insert.cc
            sql/sql_parse.cc
            sql/sql_partition_admin.cc
            sql/sql_table.cc
            sql/sql_trigger.cc
            sql/table_cache.cc
            sql/transaction.cc
            sql/wsrep_applier.cc
            sql/wsrep_binlog.cc
            sql/wsrep_hton.cc
            sql/wsrep_mysqld.cc
            sql/wsrep_sst.cc
            sql/wsrep_thd.cc
            storage/innobase/handler/ha_innodb.cc

***
    Merge conflicts resolved.
    
    ***
    Updated the WSREP_PATCH_VERSION to 53.
    
    ***
    Updated WSREP-REVISION to reflect the upstream refresh revision.
    (commit c9a1c3bd).
    
    ***
    - Adapted the rsync SST script to support SSL encryption via stunnel.
      (commit 91ea5d42 and 26bbd0d7)
    
    - In order to support encrypted SST through rsync, added '@' symbol to
      sst_method_allowed_chars_regex_pattern since it is possible to have
      '@' symbol in the SST data with the new feature.
    
      Example: rsync\000localhost:0141b8fc9fa853ac9d1d48a383de2382@127.0.0.1:13009/rsync_sst\000
    
    - Modified the script to use stunnel while transferring the keyring file
      when the server is keyring enabled.
    
    ***
    - Adapted the xtrabackup-v2 SST script to support SSL encryption by
      upstream. (commit dbdc5378 and 51fd7fc3)
    
    - Modified the parse_cnf() to consider `--loose-*` variables during
      variable normalization.
    
    Note: As part of the original commit dbdc5378, password was passed to
          innobackup via environment instead of the command line. But the
          new change had some problems in the past and had been reverted by
          PXC#460 (commit b966fc0). So, the latter behavior has been
          retained.
***
    ***
    Fixed the code differences reported by `check_src` tool.
    
    ***
    Re-recorded the test `galera.galera_defaults` for the newly introduced `wsrep_apply_waits`.
    
    ***
    Re-recorded the test `galera.galera_concurrent_ctas` due to the PXC
    deadlock warning.
    
    ***
    Update submodule pointer.
    
    ***
    Modifed the wsrep_sst_rsync script to use `verify` option for older
    versions of stunnel.
    
    - `check_for_version` and `normalize_version` bash functions are moved
      to `wsrep_sst_common`.
    
    - Typos fixed: Changes `PREFFERED` to `PREFERRED`.


Testing Done
---

Galera suite: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/186/testReport/
Failing tests:
```
1. galera_3nodes.galera_innobackupex_backup (fails only on CentOs)
```
Full Suite: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/187/console
```
Failing tests:
1. main.merge_mmap & main.merge: https://jira.percona.com/browse/PXC-3296
2. galera_3nodes.galera_innobackupex_backup (fails only on CentOs)
2. Keyring tests also fail on 5.7. Since we have removed keyring_vault from
   the default MTR test suites as it is supposed to be run separately in a
   pre-configured environment. However, they will be handled
   as part of https://jira.percona.com/browse/PXC-3297
```

ayurchen committed:
References codership/mysql-wsrep-bugs#888

Improve xtrabackup-v2 SST script compliance with native MariaDB SSL practices
and configuration.

1. Pass joiner's authentication information to donor together with address
   in State Transfer Request. This allows joiner to authenticate donor on
   connection. Previously joiner would accept data from anywhere.

2. Deprecate custom SSL configuration variables tca, tcert and tkey in favor
   of more familiar ssl-ca, ssl-cert and ssl-key. For backward compatibility
   tca, tcert and tkey are still supported.

3. Allow falling back to server-wide SSL configuration in [mysqld] if no SSL
   configuration is found in [sst] section of the config file.

4. Introduce ssl-mode variable in [sst] section that takes standard values
   and has following effects:
    - old-style SSL configuration present in [sst]: no effect
      otherwise:
    - ssl-mode=DISABLED or absent: retains old, backward compatible behavior
      and ignores any other SSL configuration
    - ssl-mode=VERIFY*: verify joiner's certificate and CN on donor,
                        verify donor's secret on joiner
                        (passed to donor via State Transfer Request)
                        BACKWARD INCOMPATIBLE BEHAVIOR
    - anything else enables new SSL configuration convetions but does not
      require verification

    ssl-mode should be set to VERIFY only in a fully upgraded cluster.

    Examples:

    [mysqld]
    ssl-cert=/path/to/cert
    ssl-key=/path/to/key
    ssl-ca=/path/to/ca

    [sst]

     -- server-wide SSL configuration is ignored, SST does not use SSL

    [mysqld]
    ssl-cert=/path/to/cert
    ssl-key=/path/to/key
    ssl-ca=/path/to/ca

    [sst]
    ssl-mode=REQUIRED

     -- use server-wide SSL configuration for SST but don't attempt to
        verify the peer identity

    [sst]
    ssl-cert=/path/to/cert
    ssl-key=/path/to/key
    ssl-ca=/path/to/ca
    ssl-mode=VERIFY_CA

     -- use SST-specific SSL configuration for SST and require verification
        on both sides

Also, pass the password to innobackup via environment instead of the command line.

Commit:
Rename galera_{load|unload}_provider.inc files

This patch changes the name of file galera_load_provider.inc and
galera_unload_provider.inc to galera_start_replication.inc and
galera_stop_replication.inc respectively.
The original file names were no longer reflecting what these files do,
after making wsrep_provider provider variable read-only.

Commit:
Merge branch '5.6-v25-fixup-provider-read-only' into 5.7-v25-fixup-provider-read-only

temeo committed:
Remove galera_var_notify_cmd dependency on system mysql client

Use MYSQL env variable provided by MTR environment to find
mysql client binary.

temeo committed:
Merge branch '5.6-v25-fix-galera_var_notify_cmd' into 5.7-v25-fix-galera_var_notify_cmd

ayurchen committed:
Merge branch 'xtrabackup-sst-improve-5.6' into xtrabackup-sst-improve-5.7

ayurchen committed:
Add SSL encryption to rsync SST via stunnel

Refs codership/mysql-wsrep-bugs#909

    1. Backport stunnel encryption from 8.0-v26 branch
    2. Deprecate `tkey`, `tcert` and `tca` options in [sst] section in favor of
       conventional `ssl-key`, `ssl-cert` and `ssl-ca`, but keep their precedence
       for backward compatibility.
    3. Default to require SSL encryption if at least SSL key and cert files are
       specified in configuration, either in [sst] or [mysqld] sections.
    4. Enable `verify*` option for stunnel on donor only if
       a. CA file is specified somewhere in the configuration
       AND
       b. it is explicitly requested in [sst] section by either specifying
       ssl-mode or CA file there.

       In this case if ssl-mode is not explicitly given, it defaults to VERIFY_CA.

       ssl-mode maps to stunnel options as follows:
       VERIFY_CA       -> verifyChain = yes
       VERIFY_IDENTITY -> verifyPeer = yes

       Example to require donor to verify joiner identity:
    ```
    [mysqld]
    ssl-cert=/path/to/cert
    ssl-key=/path/to/key
    ssl-ca=/path/to/ca

    [sst]
    ssl-mode=VERIFY_IDENTITY
    ```
    5. If SSL verification is requested, joiner verifies donor by checking the
       secret passed to donor via SST request.

ayurchen committed:
Protect Common Name from shell interpretation (allow it to contain spaces
and other symbols)

Refs codership/mysql-wsrep-bugs#913

ayurchen committed:
Merge branch 'xtrabackup-fixes-5.6' into xtrabackup-fixes-5.7

sjaakola committed:
codership/mysql-wsrep#381 crash with wsrep_recover startup and binlog purging

This pull request limits the binlog maintenance operations carried out during
wsrep_recover startup mode. Especially binlog purging should now not happen,
if server was started with wsrep_recover option

Also a new mtr test was added to try if wsrep_recover startup could run in
parallel with binlog purging

ayurchen committed:
Merge branch 'rsync-ssl-5.6' into rsync-ssl-5.7

sjaakola committed:
MDEV-25551 applying crash with tables without PK

The underlying problem with MDEV-25551 turned out to be that
transactions having changes for tables with no primary key,
were not safe to apply in parallel. This is due to excessive locking
in innodb side, and even non related row modifications could end up
in lock conflict during applying.

The fix for MDEV-25551 has disabled parallel applying for tables with no PK.
This fix depends on change for wsrep-lib, where a separate PR allows
application to modify transaction flags in wsrep-lib.

This commit has also separate mtr test for verifying that transactions
modifying a table with no primary key, will not apply in parallel.
This test is a modified version of initial test created by Gabor Orosz,
the reporterr of MDEV-25551.

ayurchen committed:
Make binary packages depend on stunnel as it is needed for rsync SST

Refs codership/mysql-wsrep-bugs#939

sjaakola committed:
Merge remote-tracking branch 'mysql-server/5.7' into 5.7-v25-mvb-933

temeo committed:
Replaced remaining DBUG_OFF directives with NDEBUG

MikkoJaakola committed:
codership/mysql-wsrep-bugs#969 porting galera_concurrent_ctas test from MariaDB 10.4

MikkoJaakola committed:
Fixes to galera_concurrent_ctas.test

MikkoJaakola committed:
codership/mysql-wsrep-bugs#969 porting galera_concurrent_ctas test from MariaDB 10.4

Fixed the result file for galera_concurrent_ctas test

temeo committed:
Recorded galera_defaults for wsrep_apply_waits

temeo committed:
Restored missing files from mysql-5.7.34

temeo committed:
codership/mysql-wsrep-bugs#972 Disable innodb_row_log_read with Galera

Disable innodb.innodb_row_log_read MTR test when run with Galera
provider enabled. The test has concurrent DDL/DML execution and
the DML will always fail in certification.

temeo committed:
Merged man pages from upstream 5.7.34

temeo committed:
Bumped wsrep patch version to 26 for 5.7.34

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/not_signed)](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1479) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1479) before we can accept your contribution.<br/>**1** out of **6** committers have signed the CLA.<br/><br/>:white_check_mark: venkatesh-prasad-v<br/>:x: sciascid<br/>:x: temeo<br/>:x: MikkoJaakola<br/>:x: sjaakola<br/>:x: ayurchen<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/percona-xtradb-cluster?pullRequest=1479) it.</sub>

kamil-holubicki commented on sql/wsrep_sst.cc:
```diff
@@ -89,18 +89,21 @@ static Regex sst_allowed_methods_regex;
+   (colon, slash, dot, underscore, square brackets, hyphen, '@' symbol).
 
-   It is the same as ^[\w:/.[\]-]+$    */
+   Note: '@' symbol is possible with encrypted rsync SST, where data can be
```
Could you extend galera_sst_allowed_methods.test, please?

venkatesh-prasad-v replied:
Sure. I'll update it.

venkatesh-prasad-v replied:
Done

kamil-holubicki commented on sql/wsrep_sst.cc:
```diff
@@ -89,18 +89,21 @@ static Regex sst_allowed_methods_regex;
    at the end of matching set (cannot be escaped when in the middle). */
 static const char *sst_method_allowed_chars_regex_pattern=
-    "^[[:alnum:]:/._<>-]+$";
+    "^[[:alnum:]:/._<>-@]+$";
```
Did you consider the problem with the hyphen mentioned in the comment? Right now it will be considered as the range between > and @, no?

venkatesh-prasad-v replied:
Yes. You're right! I remember this conversation. Thanks for your inputs. I'll modify it.

venkatesh-prasad-v replied:
Done

dutow commented on scripts/wsrep_sst_common.sh:
```diff
@@ -373,7 +384,9 @@ wsrep_gen_secret()
 }
 
->>>>>>> wsrep_5.7.34-25.26
+# Returns the absolute path from a path to a file (with a filename)
```
This isn't strictly a merge conflict, but this comment is incorrect: the function doesn't return the absolute path, it prints the absolute path and returns a status code.

venkatesh-prasad-v replied:
Fixed

kennt-percona commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -756,8 +764,40 @@ get_transfer()
+    ssl_ca=$(parse_cnf $section ssl-ca "")
+    ssl_cert=$(parse_cnf $section ssl-cert "")
+    ssl_key=$(parse_cnf $section ssl-key "")
+    if [ 0 -eq $encrypt -a -n "$ssl_cert" -a -n "$ssl_key" ]
```
this is interesting.  So if the user sets encrypt=0 (which is no encryption), then it's possible they'll get SSL encryption anyway?

venkatesh-prasad-v replied:
Adjusted the condition to
```
if [ -n "$ssl_cert" -a -n "$ssl_key" ]
```

kennt-percona commented on mysql-test/suite/galera/t/galera_sst_xtrabackup-v2_encrypt_with_server.test:
```diff
@@ -0,0 +1,24 @@
+# This test checks that if SST SSL is not explicitly donfigured mariabackup SST
+# uses server SSL configuration if present.
+# Initial SST happens via mariabackup, so there is not much to do in the body
+# of the test
```
spelling error: donfigured -> configured

venkatesh-prasad-v replied:
Done.

venkatesh-prasad-v commented:
Testing Done
---

Galera suite: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/216/testReport/
Failing tests:
```
1. galera.galera_sst_xtrabackup-v2_encrypt_options - Fixed
```
Full Suite: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/217/testReport/
```
Failing tests:
1. galera.galera_sst_xtrabackup-v2_encrypt_options - Fixed
Known failures:
1. main.merge_mmap & main.merge: https://jira.percona.com/browse/PXC-3296
2. galera_3nodes.galera_innobackupex_backup (fails only on CentOs)
2. Keyring tests also fail on 5.7. Since we have removed keyring_vault from
   the default MTR test suites as it is supposed to be run separately in a
   pre-configured environment. However, they will be handled
   as part of https://jira.percona.com/browse/PXC-3297
```

kennt-percona commented on mysql-test/suite/galera/t/galera_sst_rsync.cnf:
```diff
@@ -3,6 +3,9 @@
 [mysqld]
 wsrep_sst_method=rsync
 
+[sst]
```
since we're not using the ssl_mode, should this just be "encrypt=0"

venkatesh-prasad-v replied:
You're right! Thanks for catching this. I'll update it soon.

Commit:
Implemented PXC-3722 (Merge codership/mysql-wsrep 5.7.34 - merge with conflicts)

https://jira.percona.com/browse/PXC-3722

Merge tag 'wsrep_5.7.34-25.26' into codership_5.7.34_merge

Merge with conflicts.

Conflicts:
        cmake/wsrep.cmake
        man/mysqldumpslow.1
        mysql-test/suite/galera/r/galera_defaults.result
        scripts/wsrep_sst_common.sh
        scripts/wsrep_sst_rsync.sh
        scripts/wsrep_sst_xtrabackup-v2.sh
        sql/binlog.cc
        sql/handler.cc
        sql/log_event.cc
        sql/mdl.cc
        sql/mysqld.cc
        sql/rpl_slave.cc
        sql/sql_class.cc
        sql/sql_class.h
        sql/sql_insert.cc
        sql/sql_parse.cc
        sql/sql_partition_admin.cc
        sql/sql_table.cc
        sql/sql_trigger.cc
        sql/table_cache.cc
        sql/transaction.cc
        sql/wsrep_applier.cc
        sql/wsrep_binlog.cc
        sql/wsrep_hton.cc
        sql/wsrep_mysqld.cc
        sql/wsrep_sst.cc
        sql/wsrep_thd.cc
        storage/innobase/handler/ha_innodb.cc

Commit:
Implemented PXC-3722 (Merge codership/mysql-wsrep 5.7.34 - conflicts resolved)

https://jira.percona.com/browse/PXC-3722

Merge conflicts resolved.

***
Updated the WSREP_PATCH_VERSION to 53.

***
Updated WSREP-REVISION to reflect the upstream refresh revision.
(commit c9a1c3bd).

***
- Adapted the rsync SST script to support SSL encryption via stunnel.
  (commit 91ea5d42 and 26bbd0d7)

- In order to support encrypted SST through rsync, added '@' symbol to
  sst_method_allowed_chars_regex_pattern since it is possible to have
  '@' symbol in the SST data with the new feature.

  Example: rsync\000localhost:0141b8fc9fa853ac9d1d48a383de2382@127.0.0.1:13009/rsync_sst\000

- Modified the script to use stunnel while transferring the keyring file
  when the server is keyring enabled.

***
- Adapted the xtrabackup-v2 SST script to support SSL encryption by
  upstream. (commit dbdc5378 and 51fd7fc3)

- Modified the parse_cnf() to consider `--loose-*` variables during
  variable normalization.

Note: As part of the original commit dbdc5378, password was passed to
      innobackup via environment instead of the command line. But the
      new change had some problems in the past and had been reverted by
      PXC#460 (commit b966fc0). So, the latter behavior has been
      retained.

Commit:
Implemented PXC-3722 (Merge codership/mysql-wsrep 5.7.34 - post merge fixes)

https://jira.percona.com/browse/PXC-3722

***
Fixed the code differences reported by `check_src` tool.

***
Re-recorded the test `galera.galera_defaults` for the newly introduced `wsrep_apply_waits`.

***
Re-recorded the test `galera.galera_concurrent_ctas` due to the PXC
deadlock warning.

***
Update submodule pointer.

***
Modifed the wsrep_sst_rsync script to use `verify` option for older
versions of stunnel.

- `check_for_version` and `normalize_version` bash functions are moved
  to `wsrep_sst_common`.

- Typos fixed: Changes `PREFFERED` to `PREFERRED`.

- Added new test scenarios for validation of rsync SST requests in
  bmysql-test/suite/galera/t/galera_sst_allowed_methods.test

Commit:
Implemented PXC-3722 (Revert ssl-mode changes)

Changes done:

- Secret passing mechanism during SST introduced by Codership has been
  retained.

- SSL-MODE option introduced by Codership is removed from rsync and
  xtrabackup SST scripts. Now both these scripts support only `encrypt`
  option specified in the [sst] secion for encrypting the SST traffic.

- rsync SST script now supports `encrypt` option.

        - Setting encrypt=0 in rsync based SST disables SST encryption.

        - Setting encrypt=2/3/4 in rsync based SST is not supported and shall
          result in SST failure.

        - Setting encrypt=4 in rsync based SST enables SST encryption
          through stunnel Using ecrypt=4 requires all ssl-ca, ssl-key
          and ssl-cert for the encryption to work properly. So, absence
          of even a single file shall result in SST failure.

- While checking for the SSL files in xtrabackup based SST, the new order is
  1. sst
  2. mysqld.1 (default-suffix)
  3. mysqld

  This means that when encrypt=4 is specified, then the script first
  looks for the certificates in the [sst] secion, if not found, then in
  the mysqld.1 (with suffixes) secion and then in the mysqld section until
  the file is found.

- The functions `verify_ca_matches_cert()` and
  `verify_cert_matches_key()`` has been moved to wsrep_sst_common script
  for the reusability of the function.

- Added MTR testcases.

kennt-percona reviewed (approved):
lgtm

kamil-holubicki reviewed (approved):
lgtm
